src/main/java/seedu/address/logic/parser/TagCommandParser.java

```diff
@@ -30,51 +31,59 @@ public TagCommand parse(String args) throws ParseException {
-            }
-            checkIndexLength(indexStrings);
+        Set<Index> indexes = parseIndexes(argMultimap.getPreamble());
+        Set<Tag> tags = parseTags(argMultimap.getAllValues(PREFIX_TAG));
 
-            for (String indexStr : indexStrings) {
-                if (!indexStr.isEmpty()) {
-                    checkIndex(indexStr);
-                    indexSet.add(ParserUtil.parseIndex(indexStr.trim()));
-                }
-            }
-        } catch (NumberFormatException e) {
-            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, TagCommand.MESSAGE_USAGE));
-        }
+        return new TagCommand(new ArrayList<>(indexes), tags);
+    }
 
+    private Set<Index> parseIndexes(String input) throws ParseException {
+        Set<Index> indexSet = new HashSet<>();
+        List<String> indexStrs = List.of(input.trim().split("\\s+"));
+        checkIndexesValidity(indexStrs);
+        checkIndexLength(indexStrs);
+        addIndexToSet(indexStrs, indexSet);
+
+        return indexSet;
+    }
+
+    private Set<Tag> parseTags(List<String> tagStrings) throws ParseException {
+        Set<Tag> tags = new HashSet<>();
         for (String tagString : tagStrings) {
-            tagString = tagString.trim().toLowerCase();
-            boolean isEmpty = tagString.isEmpty();
-            boolean isTooLong = tagString.length() > Tag.MAX_CHARACTER_LENGTH;
-            if (isEmpty) {
-                throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, TagCommand.MESSAGE_USAGE));
-            } else if (isTooLong) {
-                throw new ParseException(String.format(Messages.MESSAGE_INPUT_LENGTH_EXCEEDED,
-                        Tag.MAX_CHARACTER_LENGTH));
-            }
-            tagSet.add(new Tag(tagString));
+            String trimmedTag = tagString.trim().toLowerCase();
+            validateTagName(trimmedTag);
+            tags.add(new Tag(trimmedTag));
         }
+        return tags;
+    }
 
-        return new TagCommand(new ArrayList<>(indexSet), tagSet);
+    /**
+     * Validates the given tag name to ensure it adheres to the tag naming conventions.
+     *
+     * @param tagName The name of the tag to validate.
+     * @throws ParseException If the provided tag name is invalid.
+     */
+    private void validateTagName(String tagName) throws ParseException {
+        if (!isValidTagName(tagName)) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, TagCommand.MESSAGE_USAGE));
+        }
     }
 
     private void checkIndex(String indexStr) throws ParseException {
         if (!indexStr.matches(VALIDATION_REGEX)) {
-            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
-                    TagCommand.MESSAGE_USAGE));
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, TagCommand.MESSAGE_USAGE));
```

Comment:
I'm not sure about this but are we meant to import TagCommand.MESSAGE_USAGE specifically? I remember we are supposed to list each import explicitly instead of importing groups at once but I'm not sure if it applies here

Reply:
`import seedu.address.logic.commands.TagCommand;`
This line is needed not only for the TagCommand.MESSAGE_USAGE, but in the `parse` method which returns a TagCommand. Hence having an additional `import seedu.address.logic.commands.TagCommand.MESSAGE_USAGE is redundant;`